straxen/plugins/peak_processing.py

```diff
@@ -331,98 +336,393 @@ def find_n_competing(peaks, windows, fraction):
-            s2_indices = touching_windows[p_i]
-            for s2_idx in range(s2_indices[0], s2_indices[1]):
-                s2_a = pre_s2_peaks[s2_idx]
-                if p_a['center_time'] - s2_a['center_time'] <= 0:
+        for p_i, (suspicious_peak, sigma) in enumerate(zip(peaks, sigmas)):
+            # casting_peak is the previous large peak casting shadow
+            # suspicious_peak is the suspicious peak which in shadow from casting_peak
+            indices = touching_windows[p_i]
+            for idx in range(indices[0], indices[1]):
+                casting_peak = pre_peaks[idx]
+                dt = suspicious_peak['center_time'] - casting_peak['center_time']
+                if dt <= 0:
+                    continue
+                # First we record the time difference to the nearest previous peak
+                if dt < result['nearest_dt'][p_i]:
+                    result['nearest_dt'][p_i] = dt
+                # Calculate time shadow
+                new_shadow = casting_peak['area'] * dt ** exponent
+                if pos_corr:
+                    # Calculate position shadow which is time shadow with a HalfCauchy PDF multiplier
+                    distance = distance_in_xy(suspicious_peak, casting_peak)
+                    distance = np.where(np.isnan(distance), 2 * straxen.tpc_r, distance)
+                    new_shadow *= 2 / (np.pi * sigma * (1 + (distance / sigma) ** 2))
+                # Only the previous peak with largest shadow is recorded
+                if new_shadow > result['shadow'][p_i]:
+                    result['shadow'][p_i] = new_shadow
+                    result['x'][p_i] = casting_peak['x']
+                    result['y'][p_i] = casting_peak['y']
+                    result['dt'][p_i] = suspicious_peak['center_time'] - casting_peak['center_time']
+
+
+@export
+class PeakAmbience(strax.OverlapWindowPlugin):
+    """
+    Calculate Ambience of peaks.
+    Features are the number of lonehits, small S0, S1, S2 in a time window before peaks,
+    and the number of small S2 in circle near the S2 peak in a time window.
+    References:
+        * v0.0.5 reference: xenon:xenonnt:ac:prediction:shadow_ambience
+    """
+    __version__ = '0.0.5'
+    depends_on = ('lone_hits', 'peak_basics', 'peak_positions')
+    provides = 'peak_ambience'
+    data_kind = 'peaks'
+    save_when = strax.SaveWhen.EXPLICIT
+
+    ambience_time_window_backward = straxen.URLConfig(
+        default=int(2e6),
+        type=int,
+        track=True,
+        help='Search for ambience in this time window [ns]'
+    )
+
+    ambience_divide_t = straxen.URLConfig(
+        default=False,
+        type=bool,
+        track=True,
+        help='Whether to divide area by time'
+    )
+
+    ambience_divide_r = straxen.URLConfig(
+        default=False,
+        type=bool,
+        track=True,
+        help='Whether to divide area by radius'
+    )
+
+    ambient_radius = straxen.URLConfig(
+        default=6.7,
```

Comment:
sorry @dachengx, where can I see how you optimized this parameter?  In [this note](https://xe1t-wiki.lngs.infn.it/doku.php?id=xenon:xenonnt:ac:prediction:shadow_ambience), I could not find quantitative estimation.

Reply:
It is in the [definition](https://xe1t-wiki.lngs.infn.it/doku.php?id=xenon:xenonnt:ac:prediction:shadow_ambience#definition) of Ambience.

Reply:
Which plots should I look at to understand how you optimized this 6.7 cm cut value?